## 1: Load the dataset

In this step we import the pandas library and load the raw CSV file into a table called df. 
We then look at the first few rows to get a quick idea of what the data looks like.

In [1]:
import pandas as pd

df = pd.read_csv("../data/raw/laptops_dataset_final_600.csv")
df.head(3)

,product_name,overall_rating,no_ratings,no_reviews,rating,title,review
0,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/...,4.7,"15,210",900,5,Perfect product!,"Loved it, it's my first MacBook that I earned ..."
1,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/...,4.7,"15,210",900,5,Fabulous!,Battery lasted longer than my first relationsh...
2,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/...,4.7,"15,210",900,5,Fabulous!,Such a great deal.. very happy with the perfor...


## 2: Check the size and structure

Here we check how many rows and columns the dataset has, and list the column names. 
This tells us the scale of the data we are working with.

In [2]:
print("Shape:", df.shape)
print("Columns:", df.columns.tolist())

Shape: (24113, 7)
Columns: ['product_name', 'overall_rating', 'no_ratings', 'no_reviews', 'rating', 'title', 'review']


## 3: Check data types

Each column in a dataset has a type, such as text, whole number, or decimal number. 
Here we check what type pandas has assigned to each column. 
This helps us catch columns that look numeric but are stored as text, which is a common issue.

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 24113 entries, 0 to 24112
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   product_name    24113 non-null  str    
 1   overall_rating  24113 non-null  float64
 2   no_ratings      24113 non-null  str    
 3   no_reviews      24113 non-null  str    
 4   rating          24113 non-null  int64  
 5   title           24113 non-null  str    
 6   review          24113 non-null  str    
dtypes: float64(1), int64(1), str(5)
memory usage: 1.3 MB


### Note on no_ratings and no_reviews

Both no_ratings and no_reviews are stored as text instead of numbers. 
At first glance most values look like plain numbers, for example 900. 
However, a small number of rows contain commas as thousands separators, for example 15,210. 
Because pandas requires every value in a column to fit one single type, 
the presence of even a few comma separated values forces the entire column to be stored as text. 
Both columns will need to be cleaned by removing commas and converting to numeric type 
before they can be used in any calculations.

In [11]:
print(df["no_ratings"].str.contains(",").sum(), "values in no_ratings contain a comma")
print(df["no_reviews"].str.contains(",").sum(), "values in no_reviews contain a comma")

15969 values in no_ratings contain a comma
100 values in no_reviews contain a comma


## 4: Check for missing values

Here we count how many empty or missing entries exist in each column. 
Missing data can affect analysis later, so it is important to know if any exists early.

In [4]:
df.isnull().sum()

product_name      0
overall_rating    0
no_ratings        0
no_reviews        0
rating            0
title             0
review            0
dtype: int64

## 5: Check for duplicate rows

Here we count how many rows are exact repeats of another row. 
Duplicate rows can distort our analysis, for example by making one opinion look like many.

In [5]:
df.duplicated().sum()

np.int64(7122)

## 6: Summary statistics

Here we look at basic statistics such as average, minimum, and maximum values for the numeric columns. 
This gives a quick sense of the overall pattern in the data.This only works on fully numeric columns.

In [6]:
df.describe()

,overall_rating,rating
count,24113.000000,24113.000000
mean,4.186273,4.214573
std,0.228392,1.184845
min,3.300000,1.000000
25%,4.100000,4.000000
50%,4.200000,5.000000
75%,4.300000,5.000000
max,5.000000,5.000000


## 7: Rating distribution

Here we count how many reviews fall under each star rating, from 1 to 5. 
This tells us if the dataset is balanced or if most reviews lean toward a certain rating.

In [8]:
df["rating"].value_counts()

rating
5    13972
4     5629
3     2035
1     1809
2      668
Name: count, dtype: int64

## 8: Investigate duplicate rows before deciding what to do

Before removing duplicate rows, we look at a sample of them. 
This helps us understand whether they are truly repeated entries or just short, generic reviews written by different people.

In [12]:
duplicate_rows = df[df.duplicated(keep=False)]
duplicate_rows.sort_values("review").head(20)

,product_name,overall_rating,no_ratings,no_reviews,rating,title,review
18268,Lenovo LOQ AMD Ryzen 7 Octa Core 7435HS - (24 ...,4.4,743,56,4,Worth the money,### Lenovo LOQ 15ARP9 Review\n\n**Pros:**\n\n1...
16039,Lenovo LOQ AMD Ryzen 7 Octa Core 7435HS - (24 ...,4.4,743,56,4,Worth the money,### Lenovo LOQ 15ARP9 Review\n\n**Pros:**\n\n1...
16045,Lenovo LOQ AMD Ryzen 7 Octa Core 7435HS - (24 ...,4.4,743,56,5,Worth every penny,**Gaming Laptop Review**\n\n**Pros:**\n- **Per...
18274,Lenovo LOQ AMD Ryzen 7 Octa Core 7435HS - (24 ...,4.4,743,56,5,Worth every penny,**Gaming Laptop Review**\n\n**Pros:**\n- **Per...
9550,ASUS TUF Gaming A15 AMD Ryzen 7 Octa Core 7435...,4.4,566,40,5,Super!,*In my opinion it feel pretty premium while us...
9540,ASUS TUF Gaming A15 AMD Ryzen 7 Octa Core 7435...,4.4,566,40,5,Super!,*In my opinion it feel pretty premium while us...
4820,ASUS Vivobook 15 Intel Core i3 12th Gen 1215U ...,4.3,"9,156",698,5,Worth every penny,"*Ram management is not done nicely, approx 2.5..."
6047,ASUS Vivobook 15 Intel Core i3 12th Gen 1215U ...,4.3,"9,156",698,5,Worth every penny,"*Ram management is not done nicely, approx 2.5..."
17828,ASUS Vivobook 15 Intel Core i3 12th Gen 1215U ...,4.3,"9,156",698,5,Worth every penny,"*Ram management is not done nicely, approx 2.5..."
8888,ASUS Vivobook 15 Intel Core i3 12th Gen 1215U ...,4.3,"9,156",698,5,Worth every penny,"*Ram management is not done nicely, approx 2.5..."


### Note on duplicate reviews

Investigation of the duplicate rows shows that most duplicates are long, detailed, uniquely worded reviews 
that appear multiple times under the same product, with identical text, rating, and formatting. 
It is highly unlikely that different customers independently wrote identical multi-sentence reviews. 
This suggests the duplication happened during data collection, most likely the scraper captured the 
same review more than once. 
Since these are not distinct opinions from different people, it is appropriate to remove exact 
duplicate rows during the cleaning stage using drop_duplicates.